corpus_explorer/utils.py

```diff
@@ -205,6 +207,76 @@ def generate_topic_scatter_plot(topic_coordinates, topic_proportions):
     )
 
 
+def get_topic_term_ranks(
+    docterm: List[Tuple[int]],
+    termtopic: np.ndarray,
+) -> Dict[float, Dict[int, List[int]]]:
+    """Compute term relevance rankings for all topics.
+
+    Notes
+    -----
+    The term relevance computation is described in this paper:
+    https://nlp.stanford.edu/events/illvi2014/papers/sievert-illvi2014.pdf
+
+    Term relevance is defined as an interpolation between log(P(term|topic)) and
+    log(P(term|topic) / P(term)). The interpolation is parameterized with a lambda
+    term between 0 and 1. We compute:
+
+    relevance(word, topic | lambda) =
+        lambda * log(P(term|topic)) + (1 - lambda) * log(lift(term, topic))
+
+    where lift(term, topic) = P(term|topic) / P(term).
+
+    This function computes the rankings for all lambdas in the range [0, 1], with
+    step size 0.1.
+
+    Parameters
+    ----------
+    docterm
+        A sparse document-term matrix in integer list repesentation.
+
+    termtopic
+        |topics| x |terms| np.ndarray
+        Rows are term probabilities for a single topic.
+
+    Returns
+    -------
+    A dict with the following structure:
+        {
+            lambda_value_1: {
+                topic_1: (term_id_1, term_id_2, term_id_3, ...),
+                topic_2: (term_id_x, term_id_y, term_id_z, ...),
+                ...
+            },
+            lambda_value_2: ...,
+            ...
+        }
+
+    """
+    # Convert docterm into scipy matrix and compute P(term) for all terms
+    docterm_mat = corpus2csc(docterm)
+    p_term = np.asarray(docterm_mat.sum(axis=1) / docterm_mat.sum()).squeeze()
+
+    term_ranks = {}
+
+    for lam in np.arange(0, 1.1, step=0.1):
+        lam = round(lam, 1)  # round to avoid floating point imprecision stuff
```

Comment:
This sucks, but it's what I ended up doing with the little time I had.

The problem is the following: 
I want to have these terms rankings precomputed for various values of lambda.
It would make sense to index these precomputed rankings by the lambda value itself.
Not rounding the value makes it so that sometimes for example we get `0.3000000000004` instead of `0.3`.
What I did solves the problem, but the problem was caused by the messiness of choosing floats as dictionary keys in the first place.
It would be less hacky to rely on a key that doesn't suffer from this problem.

Another idea would be to make lambda an integer between 0 and 100 lol. I don't know.

Reply:
Yeah this seems to suck. A few thoughts. 

This calculation could theoretically be done through the app, in which case no need to precompute each possible lambda step. Instead user selects a lambda and the app takes that topic and does the calculations in place. The fear here is that the doc terms matrix can be huge and would cause problems. As I understand it, dash is all server side meaning it won't need to send the doc terms matrix to the client, instead the data is loaded by the app when a user connects? I have no clue actually how dash works. What happens when ten people connect at the same time? Going to do some reading on dash. But for now I vote we precompute what we can including this.

So back to this problem. If I understand, we have a dictionary with lambda values as keys and a dictionary (containing the ranked topic terms) as values. Would it be weird to have a key called "steps" which has value that is a list of the steps ([0.0, 0.1, 0.2 ... 1.0]. And then the other keys are simply step 1 through n (where n is the number of steps). This way if we change the step size or something its handled quite well. I have never added weird metadata to a dictionary so this feels strange to me -- let me know your thoughts. 

I think ultimately as part of the app there will be a global set of options. One of which may be lambda steps. In which case having the dictionary indexed based on step index as opposed to value would be convenient.